<p style="text-align: center;"><strong>MACHINE LEARNING</strong></p>
<p style="text-align: center;"><br /> </p>
<p style="text-align: center;"><strong>GRUPO 203008067_3</strong></p>
<p style="text-align: center;"><br /> </p>
<p style="text-align: center;"><strong>FASE 4 - COMPONENTE PRACTICO - PRACTICAS SIMULADAS</strong></p>
<p style="text-align: center;"><br /> </p>
<p style="text-align: center;"><strong>PRESENTADO POR: DARWIN RAUL MERCADO DIAZ</strong></p>
<p style="text-align: center;"><br /> </p>
<p style="text-align: center;"><strong>PRESENTADO A: EDGAR ANDRES VILLABON</strong></p>
<p style="text-align: center;"><br /> <br /> </p>
<p style="text-align: center;"><strong>UNIVERSIDAD ABIERTA Y A DISTANCIA UNAD</strong></p>
<p style="text-align: center;"><br /> <br /> <br /> </p>
<p style="text-align: center;"><strong>A&Ntilde;O 2023 II PERIODO 16-04</strong></p>
<p style="text-align: center;">&nbsp;</p>

# **INTRODUCCION**

En el siguiente ejercicio se va a desarrollar la aplicación de un modelo de Machine Learning llamado Support Vector Machine, con el cual se busca automatizar el proceso de interrupción del servicio de acueducto en función del
circuito, motivo, municipio, horas, número de instalación, impacto y nombre del
contratista.

# **Acción 1: Support Vector Machine.**
Automatice el proceso de interrupción del servicio de acueducto en función del
circuito, motivo, municipio, horas, número de instalación, impacto y nombre del
contratista. Para automatizar este proceso, se tiene el problema de identificar el
segmento de circuitos que son elegibles para la suspensión del servicio para poder
dirigirse directamente a estos circuitos.

# **Desarrollo de la acción 1**

In [1]:
import pandas as pd
import numpy as np

In [2]:
# Se carga el conjunto de datos con el cual se va a trabajar
df = pd.read_excel("Anexo 1 - evaluar las interrupciones del servicio de acueducto de EPM.xlsx",header=1)
df.head()

Impacto        Circuito Servicio         Motivo  \
0    Bajo   San Cristóbal     Agua  Mantenimiento   
1    Bajo   San Cristóbal     Agua  Mantenimiento   
2    Bajo  Castilla-Bello     Agua  Mantenimiento   
3    Bajo     Aguas Frías     Agua  Modernización   
4    Bajo     Aguas Frías     Agua  Mantenimiento   

                                            Solicita Número de instalación  \
0     Unidad Operación Mantenimiento Provisión Aguas                  1242   
1     Unidad Operación Mantenimiento Provisión Aguas                   120   
2     Unidad Operación Mantenimiento Provisión Aguas                  1122   
3  Unidad Operación y  Mantenimiento Gestión Agua...                   859   
4     Unidad Operación Mantenimiento Provisión Aguas                  1564   

  Municipio                                   Dirección  \
0  Medellín   De CR96B hasta CR99B entre CL48DD y CL 49   
1  Medellín  De CR99B hasta CR100 entre CL48BB y CL 48D   
2     Bello      De CR57 hasta CR66B entre CL52 y CL 53   
3  Medellín   De CR89DB hasta CR106 entre CL30 y CL 31F   
4  Medellín   De CR100 hasta CR110 entre CL31A y CL 31F   

  Nombre del responsable   Fecha y hora esperada                  Inicio  \
0   Claudia Marcela Toro  12/17/2018 11:30:00 AM  12/17/2018 09:00:00 AM   
1   Claudia Marcela Toro  12/17/2018 03:30:00 PM  12/17/2018 01:30:00 PM   
2   Claudia Marcela Toro  12/18/2018 11:00:00 AM  12/18/2018 09:00:00 AM   
3   Claudia Marcela Toro  12/18/2018 03:00:00 PM  12/18/2018 09:00:00 AM   
4   Claudia Marcela Toro  12/18/2018 03:00:00 PM  12/18/2018 09:00:00 AM   

                      Fin Horas   Estado       Fecha de registro  \
0  12/17/2018 11:30:00 AM  2.50  ABIERTA  12/10/2018 12:00:00 AM   
1  12/17/2018 03:30:00 PM  2.00  ABIERTA  12/10/2018 12:00:00 AM   
2  12/18/2018 11:00:00 AM  2.00  ABIERTA  12/10/2018 12:00:00 AM   
3  12/18/2018 03:00:00 PM  6.00  ABIERTA  12/06/2018 12:00:00 AM   
4  12/18/2018 03:00:00 PM  6.00  ABIERTA  12/07/2018 12:00:00 AM   

                                         Explicación  \
0  Instalaciones afectadas: 1242 | Motivo: Manten...   
1  Instalaciones afectadas: 120 | Motivo: Manteni...   
2  Instalaciones afectadas: 1122 | Motivo: Manten...   
3  Instalaciones afectadas: 859 | Motivo: Empalme...   
4  Instalaciones afectadas: 1564 | Motivo: Adecua...   

                                Barrio Nombre del contratista  Tipo de aviso  
0              JUAN XXIII - LA QUIEBRA                    EPM.           IVR  
1  JUAN XXIII - LA QUIEBRA; EL SOCORRO                    EPM.      Volantes  
2        ALTAVISTA (BELLO); EL CARMELO                    EPM.           IVR  
3                         LAS VIOLETAS           REDYCO S.A.S.           IVR  
4              SAN PABLO; LAS VIOLETAS                     EPM           IVR

In [6]:
# Se toman las variables especificas con las cuales se va a desarrollar nuestro modelo
auto_v = df[["Circuito", "Motivo","Municipio", "Horas", "Número de instalación", "Impacto","Nombre del contratista ","Barrio"]]

In [7]:
# Se crea una copia del DataFrame Inicial
df=auto_v.copy()

In [8]:
# Se verifica la información de cada variable
auto_v.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 961 entries, 0 to 960
Data columns (total 8 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   Circuito                 960 non-null    object
 1   Motivo                   960 non-null    object
 2   Municipio                960 non-null    object
 3   Horas                    960 non-null    object
 4   Número de instalación    956 non-null    object
 5   Impacto                  961 non-null    object
 6   Nombre del contratista   957 non-null    object
 7   Barrio                   960 non-null    object
dtypes: object(8)
memory usage: 60.2+ KB


In [9]:
# Se eliminan los datos nulos que se tienen
auto_v = auto_v.dropna()

In [10]:
auto_v.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 953 entries, 0 to 960
Data columns (total 8 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   Circuito                 953 non-null    object
 1   Motivo                   953 non-null    object
 2   Municipio                953 non-null    object
 3   Horas                    953 non-null    object
 4   Número de instalación    953 non-null    object
 5   Impacto                  953 non-null    object
 6   Nombre del contratista   953 non-null    object
 7   Barrio                   953 non-null    object
dtypes: object(8)
memory usage: 67.0+ KB


In [24]:
auto_v.Horas = auto_v.Horas.astype(float) # Se corrige el formato de las horas a float

In [11]:
auto_v["Circuito"].unique()

array(['San Cristóbal', 'Castilla-Bello', 'Aguas Frías', 'Villa Hermosa',
       'Altos De Niquia', 'Altavista Sur', 'El Noral',
       'San Cristóbal (Elevdado)', 'Paris', 'Niquia', 'Villa Linda',
       'Palmitas', 'Volador Centro', 'Picacho', 'Paraíso',
       'Volador Norte', 'Pedregal Alto', 'Porvenir', 'Bello', 'Castilla',
       'Los Mangos', 'Berlín', 'San Antonio De Prado', 'Machado', 'Moscú',
       'Girardota', 'Piñuela', 'Aures 2', 'Pedregal Bajo', 'América',
       'Las Hamacas', 'El Totumo', 'Potrerito', 'Santo Domingo',
       'La Tolda', 'Palenque', 'Batallón', 'La Montaña', 'Popular',
       'San Esteban', 'La Cascada', 'Ayurá', 'Copacabana', 'Santa Elena',
       'Ajizal', 'Yulimar', 'Las Flores', 'Limoncito', 'Campo Valdés',
       'Doce De Octubre', 'Altavista', 'La Pastora', 'Naranjitos',
       'Versalles', 'Volador', 'Pinos', 'Pajarito', 'Villa Del Socorro',
       'Miraflores', 'Aures', 'Las Brisas', 'Asomadera', 'El Dorado',
       'Corazón', 'El Rodeo', 'Orfel

In [12]:
auto_v["Motivo"].unique() ## Se revisan los datos de las filas y encontramos un cero, se procede a eliminar

array(['Mantenimiento', 'Modernización', 'Lavado de tanque', 'Daño',
       'Programado', 'Preventivo', 'Convenios con Municipio', 0,
       'Modernización y lavado de tanque'], dtype=object)

In [13]:
reemplazo = {0: "Mantenimiento"} ## Buscamos que el cero ahora quede con valor de Mantenimiento

In [14]:
auto_v["Motivo"] = auto_v["Motivo"].replace(reemplazo)

In [15]:
auto_v["Motivo"].unique() ## Se verifica que ya no esté el valor de cero

array(['Mantenimiento', 'Modernización', 'Lavado de tanque', 'Daño',
       'Programado', 'Preventivo', 'Convenios con Municipio',
       'Modernización y lavado de tanque'], dtype=object)

In [16]:
auto_v["Municipio"].unique() # Se verifica los datos de municipio

array(['Medellín', 'Bello', 'Copacabana', 'Girardota', 'Envigado',
       'Itagüí', 'Sabaneta', 'La Estrella', 'Barbosa', 'Rionegro',
       'Caldas', 'El Retiro'], dtype=object)

In [17]:
auto_v["Impacto"].unique() # Se verifica los datos del impacto

array(['Bajo', 'Alto', 'Medio'], dtype=object)

In [18]:
auto_v["Horas"].unique() # Se verifican las horas

array(['2.50', '2.00', '6.00', '7.00', '4.00', '5.00', '3.00', '8.00',
       '10.00', '11.00', '9.00', '5.50', '12.00', '1.00', '14.00',
       '13.00', '7.50', '19.00', '24.00', '20.00', '15.00', '18.00',
       '23.00', '22.50', 9, 7, 12, 5, 14, '16.00', '22.00', '28.00',
       '35.00', '11.50', '30.00', '33.00', '36.00', '41.00', '37.00',
       '34.00', '17.00'], dtype=object)

In [19]:
auto_v["Número de instalación"].unique() # Se observa un valor de "-" en los datos, por lo tanto se procede a cambiar por NaN y luego eliminar esa fila

array([1242, 120, 1122, 859, 1564, 15975, 10, 3645, 153, 40, 4303, 7758,
       568, 181, 931, 27719, 24, 1950, 346, 8815, 37924, 26, 5882, 1686,
       503, 7617, 190, 454, 6266, 352, 21210, 218, 139, 2815, 208, 925,
       1824, 9589, 3610, 108, 12214, 3944, 1733, 579, 1769, 381, 9357,
       4047, 5029, 221, 4334, 174, 15270, 1925, 744, 3467, 14138, 360,
       4264, 612, 1566, 6372, 581, 268, 4313, 5648, 149, 180, 13828,
       10857, 8959, 35108, 10217, 9834, 1516, 13576, 114, 320, 235, 29696,
       11430, 8610, 13290, 11311, '-', 17215, 996, 277, 27, 15420, 12578,
       255, 82, 28195, 20970, 1189, 6710, 32501, 20480, 5572, 44, 20340,
       16506, 105, 36127, 4042, 22992, 1058, 10459, 1331, 156, 9945, 633,
       28557, 9913, 505, 61, 749, 117, 31, 681, 78, 327, 212, 6461, 13,
       160, 88, 197, 256, 143, 4995, 8118, 333, 1594, 39093, 5907, 33,
       1762, 3996, 1264, 953, 1376, 757, 1350, 569, 243, 1425, 175, 4268,
       154, 1850, 142, 5454, 134, 234, 1420, 328, 478, 20,

In [20]:
auto_v["Número de instalación"] = auto_v["Número de instalación"].replace("-", np.nan) # Se reemplaza del número de instalación

In [25]:
auto_v = auto_v.dropna() # Se elimina el NAN que se puso
auto_v.info() # Se verifica que se haya aplicado

<class 'pandas.core.frame.DataFrame'>
Int64Index: 952 entries, 0 to 960
Data columns (total 8 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Circuito                 952 non-null    object 
 1   Motivo                   952 non-null    object 
 2   Municipio                952 non-null    object 
 3   Horas                    952 non-null    float64
 4   Número de instalación    952 non-null    float64
 5   Impacto                  952 non-null    object 
 6   Nombre del contratista   952 non-null    object 
 7   Barrio                   952 non-null    object 
dtypes: float64(2), object(6)
memory usage: 66.9+ KB


In [26]:
# Para poder ingresar los datos al modelo, primero se utiliza OneHotEncoder, para poder codificar los datos y puedan ser convertidos a variables que soporte el modelo
from sklearn.preprocessing import OneHotEncoder
encoder = OneHotEncoder()
colsencoder = auto_v[["Circuito","Municipio","Impacto","Nombre del contratista ","Número de instalación","Motivo","Horas"]]
xatri = encoder.fit_transform(colsencoder)

In [27]:
xatri

<952x1021 sparse matrix of type '<class 'numpy.float64'>'
	with 6664 stored elements in Compressed Sparse Row format>

In [28]:
# Se usará Label Encoder para la variable y que será el Barrio
from sklearn.preprocessing import LabelEncoder
labelenco = LabelEncoder()
label = labelenco.fit_transform(auto_v["Barrio"])

In [29]:
y = label
X = xatri

In [30]:
# Se cargan las librerías que se van a utilizar
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn import svm
from sklearn.metrics import accuracy_score
from collections import Counter
from sklearn.preprocessing import StandardScaler

In [31]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.3,train_size=0.2)

In [35]:
# Se aplica el modelo de Maquinas de soporte vectorial
sv = svm.SVC(kernel="linear", C = 10, gamma=0.1)
sv.fit(X_train,y_train)
y_pred = sv.predict(X_test)
accuracy = accuracy_score(y_test,y_pred)

print(f"Prediccion del conjunto de datos de prueba {y_pred}")
print(f"Prediccion del modelo SVM no lineal {accuracy}")

Prediccion del conjunto de datos de prueba [192 284 309 492 544  30 458 531  30 485 188 188 467  30 297 467 123 385
 531 298 485 203 467 467 492  10 539 485 531 492 542 378  12 244 284 516
 485  73 467 456 467 409 244 549 244 358 292 253 492 385  30 537 256 488
  98 545 458 378 542  85 492 485 492 524 144 456 539 503 503 516 531 309
 297 203 531 385 317 544 284 531 492 284 456 492 544 544 286 409 162 420
 366 458  37 456 364  30 262 549 366 385 215 272 398 284 409 492 397 531
 284 467 309 385 459  19 244 544 539 503 153 531 366 492 374 378 467 446
 292 488 544 431 543 465 503 378 516 503 465  33 292 531 292 467 543 188
 467 309 215 261 488 410 516 465  93 385 524 436 516 393 483 358  49 253
 262 527 397 358 253 467  98 542 531 385 358 398 545 492 298 467 253 196
 244 485 467 296 467 524 492 458 531  30 192 203 374 488  27 297 347 272
 297 410 545 531 516 467 531 253 309   2 234 531 246  27 315 162 545 467
  12  30 458 492  32  27  49 527 467 366 467 539 544 456 503 244 492 410
 531 465

In [34]:
# Se muestra de manera clara los segmentos que son elegibles para la suspención
print("Conteo de las predicciones por segmentos:")
# Se crea un diccionario que contará las ocurrencias de cada segmento
predicciones_por_segmentos = Counter(labelenco.inverse_transform(y_pred))
# Se itera sobre cada segmento y su conteo para mostrar la cantidad de veces que aparece cada uno
for segmentos, conteo in predicciones_por_segmentos.items():
    print(f"{conteo} repeticiones de segmentos elegibles para la suspención del servicio y poder dirigirse directamente {segmentos} ")

Conteo de las predicciones por segmentos:
2 repeticiones de segmentos elegibles para la suspención del servicio y poder dirigirse directamente EL PROGRESO, ALTAMIRA, CORDOBA, BOSQUES DE SAN PABLO. 
9 repeticiones de segmentos elegibles para la suspención del servicio y poder dirigirse directamente LA LIBERTAD;VILLATINA;VILLA LILLIAM;EL PINAL;LOS MANGOS;ENCISO;SUCRE 
5 repeticiones de segmentos elegibles para la suspención del servicio y poder dirigirse directamente LOS BALSOS NO. 1;SAN LUCAS;EL TESORO;LOS NARANJOS (MEDELLIN);LAS LOMAS NO. 2;ALTOS DEL POBLADO;LA ASOMADERA NO. 2 
18 repeticiones de segmentos elegibles para la suspención del servicio y poder dirigirse directamente SURAMERICANA 
10 repeticiones de segmentos elegibles para la suspención del servicio y poder dirigirse directamente ZONA INDUSTRIAL 3 (BELLO);PANAMERICANO;CIUDAD NIQUIA;LA CIUDADELA DEL NORTE;LAS GRANJAS (BELLO);ZONA INDUSTRIAL 2 (BELLO);CENTRO (BELLO);PRADO (BELLO);ZONA INDUSTRIAL 1 (BELLO);LA ESTACION;MANCHEST

# **Análisis de resultados**

- En los resultados obtenido se pudo observar que la predicción del modelo fue del 19%, porcentaje que es bastante bajo según lo que se espera para este tipo de predicciones.

- En el resultado obtenido al mostrar las predicciones de nuestro modelo, se pudo obtener los segmentos de circuitos que son elegibles para la suspención del servicio, donde se encontró que el circuito de San cristobal es donde mayor frecuencia de elegibilidad hay con 26, se puede decir que es el circuito que mas afecciones va a tener.

- Con este modelo aplicado se puede observar que las variables utilizadas no sean las ideales para las maquinas de soporte vectorial, por lo tanto se debería de intentar utilizar otras variables con las cuales se pueda obtener un resultado mucho mejor a partir de mejorar las predicciones del modelo obtenido.

# **CONCLUCIONES**

- Se pudo aplicar de manera correcta el modelo de maquinas de soporte vectorial, utilizando las varibles mencionadas al principio del ejercicio.

- Se debe de analizar si se pueden cambiar las variables utilizadas para mejorar la predicción del modelo, dado que las que se tienen dan una probabilidad del 19% que es una probabilidad baja, con lo cuál no se pueden tener resultados completamente eficientes.

- Se aprendío a utilizar las maquinas de soporte vectorial y su aplicación en problemas de clasificación y regresión.

# **REFERENCIAS**

- Kane, F. (2017). Hands-On Data Science and Python Machine Learning. Packt Publishing. (pp. 198-204). https://bibliotecavirtual.unad.edu.co/login?url=https://search.ebscohost.com/login.aspx?direct=true&db=nlebk&AN=1566405&lang=es&site=eds-live&scope=site&ebv=EB&ppid=pp_198

- Raschka, S., & Mirjalili, V. (2017). Python Machine Learning - Second Edition: Vol. 2nd ed. Packt Publishing. (pp. 79-90). https://bibliotecavirtual.unad.edu.co/login?url=https://search.ebscohost.com/login.aspx?direct=true&db=nlebk&AN=1606531&lang=es&site=eds-live&scope=site&ebv=EB&ppid=pp_79